### PR TITLE
[SPARK-7586][ML][doc] Add docs of Word2Vec in ml package

### PR DESCRIPTION
**Pull request by yinxusen**
CC @jkbradley.

JIRA [issue](https://issues.apache.org/jira/browse/SPARK-7586).

### CODE REVIEW

File: docs/ml-features.md [hunk 1]
```diff
@@ -106,6 +106,84 @@ for features_label in featurized.select("features", "label").take(3):
...
+`Word2Vec` is an `Estimator` which takes sequences of words that represents documents and trains a `Word2VecModel`. The model is a `Map(String, Vector)` essentially, which maps each word to an unique fix-sized vector. The `Word2VecModel` transforms each documents into a vector using the average of all words in the document, which aims to other computations of documents such as similarity calculation consequencely. Please refer to the [MLlib user guide on Word2Vec](mllib-feature-extraction.html#Word2Vec) for more details on Word2Vec.
+
+Word2Vec is implemented in [Word2Vec](api/scala/index.html#org.apache.spark.ml.feature.Word2Vec). In the following code segment, we start with a set of documents, each of them is represented as a sequence of words. For each document, we transform it into a feature vector. This feature vector could then be passed to a learning algorithm.
+
+<div class="codetabs">
+<div data-lang="scala" markdown="1">
+{% highlight scala %}
+import org.apache.spark.ml.feature.Word2Vec
+
+val documentDF = sqlContext.createDataFrame(Seq(
```
Comment (jkbradley) [inline docs/ml-features.md]: Add comment in line above:
```
Input data: Each row is a bag of words from a sentence or document.
```
(Please add to other examples too.)

File: docs/ml-features.md [hunk 2]
```diff
@@ -106,6 +106,84 @@ for features_label in featurized.select("features", "label").take(3):
...
+  "Hi I heard about Spark".split(" "),
+  "I wish Java could use case classes".split(" "),
+  "Logistic regression models are neat".split(" ")
+)).map(Tuple1.apply).toDF("text")
+
+val word2Vec = new Word2Vec.setInputCol("text").setOutputCol("result").setVectorSize(3)
```
Comment (jkbradley) [inline docs/ml-features.md]: Add comment in line above:
```
Learn a mapping from words to Vectors
```
(Please add to other examples too.)

File: docs/ml-features.md [hunk 3]
```diff
@@ -106,6 +106,84 @@ for features_label in featurized.select("features", "label").take(3):
...
+val model = word2Vec.fit(documentDF)
+val result = model.transform(documentDF).select("result").take(3).foreach(println)
```
Comment (jkbradley) [inline docs/ml-features.md]: Please split into 2 lines, one creating the transformed dataframe + one selecting data and printing it. (Same for other examples too)

File: docs/ml-features.md [hunk 4]
```diff
@@ -106,6 +106,84 @@ for features_label in featurized.select("features", "label").take(3):
...
+
+JavaSparkContext jsc = ...
+SQLContext sqlContext = ...
+JavaRDD<Row> jrdd = jsc.parallelize(Lists.newArrayList(
+  RowFactory.create(Lists.newArrayList("Hi I heard about Spark".split(" "))),
+  RowFactory.create(Lists.newArrayList("I wish Java could use case classes".split(" "))),
+  RowFactory.create(Lists.newArrayList("Logistic regression models are neat".split(" ")))
+));
+StructType schema = new StructType(new StructField[]{
+  new StructField("text", new ArrayType(StringType$.MODULE$, true), false, Metadata.empty())
```
Comment (jkbradley) [inline docs/ml-features.md]: Use ```DataTypes.StringType()``` instead of ```StringType$.MODULE$``` for Java

File: docs/ml-features.md [hunk 5]
```diff
@@ -106,6 +106,84 @@ for features_label in featurized.select("features", "label").take(3):
...
+<div data-lang="python" markdown="1">
+{% highlight python %}
+from pyspark.ml.feature import Word2Vec
+
+documentDF = sqlContext.createDataFrame([
+  ("Hi I heard about Spark".split(" "), ),
+  ("I wish Java could use case classes".split(" "), ),
+  ("Logistic regression models are neat".split(" "), )
+], ["text"])
+word2Vec = Word2Vec(vectorSize = 3, minCount = 0, inputCol = "text", outputCol = "result")
```
Comment (jkbradley) [inline docs/ml-features.md]: Python style: no spaces around "=" in argument list:
```
word2Vec = Word2Vec(vectorSize=3, minCount=0, inputCol="text", outputCol="result")
```

File: mllib/src/test/java/org/apache/spark/ml/feature/JavaWord2VecSuite.java [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+  public void tearDown() {
+    jsc.stop();
+    jsc = null;
+  }
+
+  @Test
+  public void testJavaWord2Vec() {
+    JavaRDD<Row> jrdd = jsc.parallelize(Lists.newArrayList(
+      RowFactory.create(Lists.newArrayList("Hi I heard about Spark".split(" ")),
+        Vectors.dense(0.017877750098705292, -0.018388677015900613, -0.01183266043663025)),
```
Comment (jkbradley) [inline mllib/src/test/java/org/apache/spark/ml/feature/JavaWord2VecSuite.java]: Let's not check correctness in Java.  Please just make sure that each element of the API can be called successfully.

File: mllib/src/test/java/org/apache/spark/ml/feature/JavaWord2VecSuite.java [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+      RowFactory.create(Lists.newArrayList("I wish Java could use case classes".split(" ")),
+        Vectors.dense(0.0038498884865215844, -0.07299017374004636, 0.010990704176947474)),
+      RowFactory.create(Lists.newArrayList("Logistic regression models are neat".split(" ")),
+        Vectors.dense(0.017819208838045598, -0.006920230574905872, 0.022744188457727434))
+    ));
+    StructType schema = new StructType(new StructField[]{
+      new StructField("text", new ArrayType(StringType$.MODULE$, true), false, Metadata.empty()),
```
Comment (jkbradley) [inline mllib/src/test/java/org/apache/spark/ml/feature/JavaWord2VecSuite.java]: Use ```DataTypes.StringType()``` instead of ```StringType$.MODULE$```

**Comment by jkbradley** [comment]
@yinxusen  I tried the Scala example but got this error:
```
scala> val documentDF = sqlContext.createDataFrame(Seq(
     |   "Hi I heard about Spark".split(" "),
     |   "I wish Java could use case classes".split(" "),
     |   "Logistic regression models are neat".split(" ")
     | )).map(Tuple1.apply).toDF("text")
<console>:20: error: overloaded method value createDataFrame with alternatives:
  [A <: Product](data: Seq[A])(implicit evidence$4: reflect.runtime.universe.TypeTag[A])org.apache.spark.sql.DataFrame <and>
  [A <: Product](rdd: org.apache.spark.rdd.RDD[A])(implicit evidence$3: reflect.runtime.universe.TypeTag[A])org.apache.spark.sql.DataFrame
 cannot be applied to (Seq[Array[String]])
       val documentDF = sqlContext.createDataFrame(Seq(
```
Did these all run for you?

The Java file needs the Apache license at the top.

**Comment by jkbradley** [comment]
@yinxusen  This needs the Apache license at the top of all files.  (That is causing the RAT test failure.)

**Comment by jkbradley** [comment]
@yinxusen Except for the RAT test issue, this LGTM.  You may encounter merge issues from other updates to ml-features.md, but they should be easy to fix.

**Comment by yinxusen** [comment]
@jkbradley Sorry for forgetting the Apache license again... fix it now.

**Comment by yinxusen** [comment]
@jkbradley Seems it has no merging problem, for now.

**Comment by yinxusen** [comment]
@jkbradley Are there more concrete error infos for Mima error? I search from the console output of the test, but cannot locate the error.

**Comment by jkbradley** [comment]
@yinxusen It's not actually a MIMA error; that's a bug in the script reporting on the tests.  It looks to me like Jenkins being flaky.

**Comment by jkbradley** [comment]
LGTM merging into master and branch-1.4
@yinxusen Thank you!

**Pull request closed without merging**